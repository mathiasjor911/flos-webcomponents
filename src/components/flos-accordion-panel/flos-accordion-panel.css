.panel-heading {
  background: transparent;
  border-bottom: 1px solid #cad7e0;
  display: block;
  margin: 0;
  padding: 16px;
  position: relative;
  transition: all 0.2s, outline 0s;
}
.panel-heading:hover {
  background-color: #e4eaef;
  border-bottom: 1px solid #53759b;
}
.panel-heading:focus {
  outline: 3px solid #63aedb;
}
.panel-heading[aria-expanded='true'] {
  background-color: #eaf1f9;
  border-bottom: none;
}
.panel-heading svg {
  transition: all 0.2s;
  position: absolute;
  right: 16px;
  top: 16px;
}
.panel-heading[aria-expanded='true'] svg {
  transform: rotate(-180deg);
}
.panel-heading h4 {
  font-family: 'Mukta', sans-serif;
  font-size: 20px;
  color: #53759b;
  font-weight: 500;
  margin: 0;
}
.panel-heading p {
  font-family: 'Mukta', sans-serif;
  color: #131947;
  margin-bottom: 0;
  margin-top: 5px;
  font-weight: 300;
}
.panel-body {
  display: none;
  background-color: #eaf1f9;
  padding: 20px;
}
.panel-body p {
	font-family: 'Mukta', sans-serif;
  margin: 0 0 16px;
  color: #131947;
  font-weight: 300;
}
.panel-heading[aria-expanded='true'] + .panel-body {
  display: block;
}

:host {
  all: inherit;
}
