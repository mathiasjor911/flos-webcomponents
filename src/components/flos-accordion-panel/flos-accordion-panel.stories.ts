export default {
  title: 'Flos/Accordion',
  // More on argTypes: https://storybook.js.org/docs/html/api/argtypes
  argTypes: {
    expanded: { type: 'boolean',control: 'boolean', description: 'Boolean that controls wheter or not the panel is open' },
    heading: { type: 'string', control: 'text', description: 'Sets the title of the Accordion Panel' },
    subtitle: { type: 'string', control: 'text', description: 'Sets the subtitle of the Accordion Panel'},
    panelId: {type: 'string', description: 'Uniqe ID for each accordion panel, passed by the Accordion Wrapper '},
    toggle: {type: '() => void', description: 'Prop passed by the Accordion wrapper to control panel toggle'}
  },
};

const loremIpsum =
  'Lorem ipsum dolor sit amet, consectetur adipiscing elit. Duis ipsum risus, sagittis in nisl eget, mattis luctus velit. Suspendisse sed auctor libero, maximus mattis nibh. Pellentesque gravida, est a dignissim convallis, erat tellus hendrerit turpis, vel fermentum massa nibh mattis nulla. Suspendisse in volutpat turpis.';

const PanelTemplate = ({...args}) => {
  return `
  <flos-accordion-panel heading="${args.heading}" ${args.expanded ? 'expanded' : ''} subtitle="${args.subtitle}">
    <p>${loremIpsum}</p>
  </flos-accordion-panel>`;
};

export const AccordionPanel = PanelTemplate.bind({})
AccordionPanel.args = {
  heading: 'Hello World',
  subtitle: 'How are you doing',
  expanded: false,
}






