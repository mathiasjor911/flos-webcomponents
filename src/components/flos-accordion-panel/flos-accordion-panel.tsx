import { Component, Prop, h, Host } from '@stencil/core';

@Component({
  tag: 'flos-accordion-panel',
  styleUrl: 'flos-accordion-panel.css',
})
/** Accordion Panel should only be used as a child of Accordion */
export class AccordionPanel {
  /**  Sets the title of the Accordion Panel */
  @Prop() heading: string;
  /** Sets the subtitle of the Accordion Panel */
  @Prop() subtitle?: string;
  /** Boolean that controls wheter or not the panel is open */
  @Prop({mutable:true, reflect: true}) expanded: boolean;
  /** Uniqe ID for each accordion panel */
  @Prop({ reflect: true }) panelId: string;
  /** Prop passed by the Accordion wrapper to control panel toggle */
  @Prop() toggle: () => void;

  render() {
    return (
      <Host>
        <div
          class="panel-heading"
          role="button"
          id={`panel-heading-${this.panelId}`}
          onClick={() => {
            this.toggle()
          }}
          aria-expanded={this.expanded ? 'true' : 'false'}
          aria-controls={`panel-${this.panelId}`}
          tabIndex={0}
        >
          <span>
            <h4>{this.heading}</h4>
            {this.subtitle && <p>{this.subtitle}</p>}
            <svg
              xmlns="http://www.w3.org/2000/svg"
              width="24"
              height="24"
              viewBox="0 0 24 24"
              fill="none"
              stroke="#131947"
              stroke-width="2"
              stroke-linecap="round"
              stroke-linejoin="round"
            >
              <polyline points="6 9 12 15 18 9"></polyline>
            </svg>
          </span>
        </div>
        <div id={`panel-${this.panelId}`} class={'panel-body'} aria-labelledby={`panel-heading-${this.panelId}`}>
          <slot />
        </div>
      </Host>
    );
  }
}
