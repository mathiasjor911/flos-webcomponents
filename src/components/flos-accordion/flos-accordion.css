.accordion-heading {
  background: transparent;
  border-bottom: 1px solid #cad7e0;
  display: block;
  margin: 0;
  padding: 16px;
  position: relative;
  transition: all 0.2s, outline 0s;
}
.accordion-heading:hover {
  background-color: #e4eaef;
  border-bottom: 1px solid #53759b;
}
.accordion-heading:focus {
  outline: 3px solid #63aedb;
}
.accordion-heading[aria-expanded='true'] {
  background-color: #eaf1f9;
  border-bottom: none;
}
.accordion-heading svg {
  transition: all 0.2s;
  position: absolute;
  right: 16px;
  top: 16px;
}
.accordion-heading[aria-expanded='true'] svg {
  transform: rotate(-180deg);
}
.accordion-heading h4 {
  font-family: 'Mukta', sans-serif;
  font-size: 20px;
  color: #53759b;
  font-weight: 500;
  margin: 0;
}
.accordion-heading p {
  font-family: 'Mukta', sans-serif;
  color: #131947;
  margin-bottom: 0;
  margin-top: 5px;
  font-weight: 300;
}
.accordion-panel {
  display: none;
  background-color: #eaf1f9;
  padding: 20px;
}
.accordion-panel p {
	font-family: 'Mukta', sans-serif;
  margin: 0 0 16px;
  color: #131947;
  font-weight: 300;
}
.accordion-heading[aria-expanded='true'] + .accordion-panel {
  display: block;
}

:host {
  all: inherit;
}
