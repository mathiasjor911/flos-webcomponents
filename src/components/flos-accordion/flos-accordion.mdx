The Accordion consists of two elements. A `flos-accordion` component, which acts as a wrapper for the `flos-accordion-panel`'s
