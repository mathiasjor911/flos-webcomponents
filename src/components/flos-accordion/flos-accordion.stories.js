import flosAccordion from './flos-accordion.mdx';

export default {
  title: 'Flos/Accordion',
  argTypes: {
    activeIndex: { type: 'string', control: 'text', description: 'Prop for controlling which panel is expanded.' },
  },
};

const loremIpsum =
  'Lorem ipsum dolor sit amet, consectetur adipiscing elit. Duis ipsum risus, sagittis in nisl eget, mattis luctus velit. Suspendisse sed auctor libero, maximus mattis nibh. Pellentesque gravida, est a dignissim convallis, erat tellus hendrerit turpis, vel fermentum massa nibh mattis nulla. Suspendisse in volutpat turpis.';

const AccordionTemplate = ({ ...args }) => {
  return `
  <flos-accordion active-index="${args.activeIndex ? args.activeIndex : ''}">
  <flos-accordion-panel heading="Udvidet elforsikring"
  subtitle="Erstatter din elektronik til nyværdi, indtil tingene er fire år gamle.">
    <p>${loremIpsum}</p>
  </flos-accordion-panel>
  </flos-accordion>
  `;
};
export const AccordionWithPanel = AccordionTemplate.bind({});

const AccordionActiveIndexTemplate = ({ ...args }) => {
  return `
  <flos-accordion active-index="${args.activeIndex ? args.activeIndex : ''}">
  <flos-accordion-panel heading="Udvidet elforsikring"
  subtitle="Erstatter din elektronik til nyværdi, indtil tingene er fire år gamle.">
    <p>${loremIpsum}</p>
  </flos-accordion-panel>
	<flos-accordion-panel heading="Udvidet elforsikring"
  subtitle="Erstatter din elektronik til nyværdi, indtil tingene er fire år gamle.">
    <p>${loremIpsum}</p>
  </flos-accordion-panel>
	<flos-accordion-panel heading="Udvidet elforsikring"
  subtitle="Erstatter din elektronik til nyværdi, indtil tingene er fire år gamle.">
    <p>${loremIpsum}</p>
  </flos-accordion-panel>
  </flos-accordion>
  `;
};
export const AccordionWithActiveIndexTemplate = AccordionActiveIndexTemplate.bind({});
AccordionWithActiveIndexTemplate.args = {
  activeIndex: '1',
};
