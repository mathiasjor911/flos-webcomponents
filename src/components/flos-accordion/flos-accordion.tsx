import { Component, Prop, h, State, Element, Watch, Listen } from '@stencil/core';
@Component({
  tag: 'flos-accordion',
  shadow: true,
})
export class Accordion {
  @Element() host: HTMLElement;
  /** Prop for controlling which panel is expanded.*/
  @Prop({ mutable: true, reflect: true }) activeIndex?: string;
  /** Internal state for controlling the current open panel */
  @State() activePanel: string;

  @Watch('activePanel')
  watchActivePanel(newValue: string) {
    this.expandActivePanel(newValue);
  }
  // Bliver kørt når komponenten bliver declared
  connectedCallback() {
    this.allPanels().forEach((panel, index) => {
      panel.panelId = index.toString();
      panel.toggle = () => this.updateState(panel.panelId);
    });
  }
  // Bliver kørt før komponentens første render
  componentWillRender() {
    if (this.activeIndex && this.activeIndex !== this.activePanel) {
      this.activePanel = this.activeIndex;
    }
  }

  // Eventlistener til keyboard events
  @Listen('keydown', { capture: true })
  onKeyDown(e: any) {
    let eventTarget = [e.target];
    let panel = eventTarget[0].parentElement;

    switch (e.key) {
      case 'Enter':
      case ' ':
        this.updateState(panel.panelId);
    }
  }

  private allPanels = () => Array.from(this.host.querySelectorAll('flos-accordion-panel'));

  private updateState(panelId: string) {
    const newActivePanel = this.activePanel === panelId ? '' : panelId;
    this.activePanel = newActivePanel;
    this.activeIndex = this.activePanel;
  }

  private expandActivePanel(activePanel) {
    this.allPanels().forEach(panel => {
      panel.expanded = panel.panelId === activePanel;
    });
  }

  render() {
    return <slot />
  }
}
