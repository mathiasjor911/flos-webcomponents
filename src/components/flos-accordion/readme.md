# flos-accordion-panel



<!-- Auto Generated Below -->


----------------------------------------------

*Built with [StencilJS](https://stenciljs.com/)*
