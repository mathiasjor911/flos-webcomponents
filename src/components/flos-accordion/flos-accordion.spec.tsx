import { newSpecPage } from '@stencil/core/testing';
import { h } from '@stencil/core';
import { Accordion } from './flos-accordion';
import { AccordionPanel } from '../flos-accordion-panel/flos-accordion-panel';

describe('Accordion with panels', () => {
  it('Renders correctly w/o subtitle', async () => {
    const page = await newSpecPage({
      components: [Accordion, AccordionPanel],
      template: () => (
        <flos-accordion>
          <flos-accordion-panel heading="Udvidet elforsikring" id="test">
            <p>Den udvidede elektronikforsikring gør individuelle elektronikforsikringer overflødige.</p>
          </flos-accordion-panel>
        </flos-accordion>
      ),
    });
    expect(page.root).toMatchSnapshot();
  });

  it('Renders correctly w/ subtitle', async () => {
    const page = await newSpecPage({
      components: [Accordion, AccordionPanel],
      template: () => (
        <flos-accordion>
          <flos-accordion-panel id="test" heading="Udvidet elforsikring" subtitle="Erstatter din elektronik til nyværdi, indtil tingene er fire år gamle.">
            <p>Den udvidede elektronikforsikring gør individuelle elektronikforsikringer overflødige.</p>
          </flos-accordion-panel>
        </flos-accordion>
      ),
    });
    expect(page.root).toMatchSnapshot();
  });

  it('Expands panel when active index is defined', async () => {
    const page = await newSpecPage({
      components: [Accordion, AccordionPanel],
      template: () => (
        <flos-accordion activeIndex="0">
          <flos-accordion-panel id="test" heading="Udvidet elforsikring" subtitle="Erstatter din elektronik til nyværdi, indtil tingene er fire år gamle.">
            <p>Den udvidede elektronikforsikring gør individuelle elektronikforsikringer overflødige.</p>
          </flos-accordion-panel>
        </flos-accordion>
      ),
    });

    const panel = page.doc.querySelector('flos-accordion-panel');
    expect(panel.expanded).toBe(true);
  });

  it('Expands panel on click', async () => {
    const page = await newSpecPage({
      components: [Accordion, AccordionPanel],
      template: () => (
        <flos-accordion>
          <flos-accordion-panel id="test" heading="Udvidet elforsikring" subtitle="Erstatter din elektronik til nyværdi, indtil tingene er fire år gamle.">
            <p>Den udvidede elektronikforsikring gør individuelle elektronikforsikringer overflødige.</p>
          </flos-accordion-panel>
        </flos-accordion>
      ),
    });

    const panel = page.doc.querySelector('flos-accordion-panel');
    const panelHeading = page.doc.querySelector('.panel-heading') as HTMLElement;

    panelHeading.click();

    await page.waitForChanges()
    expect(panel.expanded).toBe(true);
  });

  it('Closes open panel when a new panel is clicked', async () => {
    const page = await newSpecPage({
      components: [Accordion, AccordionPanel],
      template: () => (
        <flos-accordion activeIndex='0'>
          <flos-accordion-panel id="test2" heading="Udvidet elforsikring" subtitle="Erstatter din elektronik til nyværdi, indtil tingene er fire år gamle.">
            <p>Den udvidede elektronikforsikring gør individuelle elektronikforsikringer overflødige.</p>
          </flos-accordion-panel>
          <flos-accordion-panel id="test1" heading="Udvidet elforsikring" subtitle="Erstatter din elektronik til nyværdi, indtil tingene er fire år gamle.">
            <p>Den udvidede elektronikforsikring gør individuelle elektronikforsikringer overflødige.</p>
          </flos-accordion-panel>
        </flos-accordion>
      ),
    });

    const panel1 = page.doc.querySelectorAll('flos-accordion-panel')[0];
    const panel2 = page.doc.querySelectorAll('flos-accordion-panel')[1];

    const panel2Heading = page.doc.querySelectorAll('.panel-heading')[1]as HTMLElement;

    panel2Heading.click();

    await page.waitForChanges()
    expect(panel1.expanded).toBe(false);
    expect(panel2.expanded).toBe(true);
  });
});
