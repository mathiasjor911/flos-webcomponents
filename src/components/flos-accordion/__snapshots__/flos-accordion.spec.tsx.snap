// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Accordion with panels Renders correctly w/ subtitle 1`] = `
<flos-accordion>
  <mock:shadow-root>
    <slot></slot>
  </mock:shadow-root>
  <flos-accordion-panel id="test" panel-id="0">
    <!---->
    <div aria-controls="panel-0" aria-expanded="false" class="panel-heading" id="panel-heading-0" role="button" tabindex="0">
      <span>
        <h4>
          Udvidet elforsikring
        </h4>
        <p>
          Erstatter din elektronik til nyværdi, indtil tingene er fire år gamle.
        </p>
        <svg fill="none" height="24" stroke="#131947" stroke-linecap="round" stroke-linejoin="round" stroke-width="2" viewBox="0 0 24 24" width="24" xmlns="http://www.w3.org/2000/svg">
          <polyline points="6 9 12 15 18 9"></polyline>
        </svg>
      </span>
    </div>
    <div aria-labelledby="panel-heading-0" class="panel-body" id="panel-0">
      <p>
        Den udvidede elektronikforsikring gør individuelle elektronikforsikringer overflødige.
      </p>
    </div>
  </flos-accordion-panel>
</flos-accordion>
`;

exports[`Accordion with panels Renders correctly w/o subtitle 1`] = `
<flos-accordion>
  <mock:shadow-root>
    <slot></slot>
  </mock:shadow-root>
  <flos-accordion-panel id="test" panel-id="0">
    <!---->
    <div aria-controls="panel-0" aria-expanded="false" class="panel-heading" id="panel-heading-0" role="button" tabindex="0">
      <span>
        <h4>
          Udvidet elforsikring
        </h4>
        <svg fill="none" height="24" stroke="#131947" stroke-linecap="round" stroke-linejoin="round" stroke-width="2" viewBox="0 0 24 24" width="24" xmlns="http://www.w3.org/2000/svg">
          <polyline points="6 9 12 15 18 9"></polyline>
        </svg>
      </span>
    </div>
    <div aria-labelledby="panel-heading-0" class="panel-body" id="panel-0">
      <p>
        Den udvidede elektronikforsikring gør individuelle elektronikforsikringer overflødige.
      </p>
    </div>
  </flos-accordion-panel>
</flos-accordion>
`;
